src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+//   auto result = machine.withState<Active>([](Active& a) {
+//     return a.computeSomething();
+//   });
+//   machine.transitionTo<Closed>();
+//
+// =============================================================================
+// QUICK START
+// =============================================================================
+//
+// Define your state types (add NAME for introspection):
+//
+//   struct Readable {
+//     static constexpr kj::StringPtr NAME = "readable"_kj;
+//     kj::Own<Source> source;
+//   };
+//   struct Closed { static constexpr kj::StringPtr NAME = "closed"_kj; };
+//   struct Errored {
+//     static constexpr kj::StringPtr NAME = "errored"_kj;
+//     jsg::Value error;
+//   };
+//
+// Basic state machine with safe access:
+//
+//   StateMachine<Readable, Closed, Errored> state;
+//   state.transitionTo<Readable>(...);
+//
+//   // RECOMMENDED: Use withState() for safe scoped access
+//   state.withState<Readable>([](Readable& r) {
+//     r.source->read();  // Safe - transitions blocked during callback
+//   });
+//
+//   // Or with a return value
+//   auto size = state.withState<Readable>([](Readable& r) {
+//     return r.source->size();
+//   });  // Returns kj::Maybe<size_t>
+//
+// Stream-like state machine (common pattern in workerd):
+//
+//   ComposableStateMachine<
+//       TerminalStates<Closed, Errored>,
+//       ErrorState<Errored>,
+//       ActiveState<Readable>,
+//       PendingStates<Closed, Errored>,
+//       Readable, Closed, Errored
+//   > state;
+//
+//   state.transitionTo<Readable>(...);
+//
+//   // Safe access with whenActive()
+//   state.whenActive([](Readable& r) {
+//     r.source->doSomething();  // Transitions blocked
+//   });
+//
+//   // Error checking
+//   if (state.isErrored()) { ... }
+//   KJ_IF_SOME(err, state.tryGetError()) { ... }
+//
+//   // Deferred transitions during operations
+//   state.beginOperation();
+//   state.deferTransitionTo<Closed>();  // Deferred until operation ends
+//   state.endOperation();               // Now transitions to Closed
+//
+//   // Terminal enforcement
+//   state.transitionTo<Closed>();
+//   state.transitionTo<Readable>(...);  // FAILS - can't leave terminal state
+//
+// =============================================================================
+// MIGRATION GUIDE: From kj::OneOf to StateMachine
+// =============================================================================
+//
+// This section describes how to migrate existing kj::OneOf state management
+// to use these StateMachine utilities.
+//
+// STEP 1: Add NAME constants to state types
+// -----------------------------------------
+// StateMachine provides currentStateName() for debugging. Add NAME to states:
+//
+//   // Before:
+//   struct Closed {};
+//
+//   // After:
+//   struct Closed {
+//     static constexpr kj::StringPtr NAME = "Closed"_kj;
+//   };
+//
+// Note: If using ErrorState<T>, T must be a struct (not a type alias).
```

Comment:
The "not a type alias" part is hard for me to believe. Given

```cpp
struct Errored { static constexpr kj::StringPtr NAME = "errored"_kj; };
using Errored_ = Errored;
```

`ErrorState<Errored_>` really won't work?